kernel/memory_structs/src/lib.rs

```diff
@@ -266,87 +315,145 @@ macro_rules! implement_page_frame {
+                }
+
+                #[doc = "Returns a 2MiB huge`" $TypeName "` containing the given `" $address "`."]
+                pub const fn containing_address_2mb(addr: $address) -> $TypeName<Page2MiB> {
+                    $TypeName {
+                        number: addr.value() / (PAGE_SIZE * Page2MiB::CONTAINED_4K) * Page2MiB::CONTAINED_4K,
+                        size: PhantomData::<Page2MiB>,
                     }
                 }
 
-                #[doc = "Returns a new `" $TypeName "` that is aligned up from this \
-                    `" $TypeName "` to the nearest multiple of `alignment_4k_pages`."]
-                #[doc(alias = "next_multiple_of")]
-                pub const fn align_up(&self, alignment_4k_pages: usize) -> $TypeName {
+                #[doc = "Returns a 1GiB huge `" $TypeName "` containing the given `" $address "`."]
+                pub const fn containing_address_1gb(addr: $address) -> $TypeName<Page1GiB> {
                     $TypeName {
-                        number: self.number.next_multiple_of(alignment_4k_pages)
+                        number: addr.value() / (PAGE_SIZE * Page1GiB::CONTAINED_4K) * Page1GiB::CONTAINED_4K,
+                        size: PhantomData::<Page1GiB>,
                     }
                 }
+
+                #[doc = "Returns the size of this `" $TypeName "`."]
+                pub const fn page_size(&self) -> MemChunkSize {
+                    P::SIZE
+                }
             }
-            impl fmt::Debug for $TypeName {
+            impl<P: 'static + PageSize> fmt::Debug for $TypeName<P> {
                 fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
                     write!(f, concat!(stringify!($TypeName), "(", $prefix, "{:#X})"), self.start_address())
                 }
             }
-            impl Add<usize> for $TypeName {
-                type Output = $TypeName;
-                fn add(self, rhs: usize) -> $TypeName {
+            impl<P: PageSize> Add<usize> for $TypeName<P> {
+                type Output = $TypeName<P>;
+                fn add(self, rhs: usize) -> $TypeName<P> {
                     // cannot exceed max page number (which is also max frame number)
                     $TypeName {
-                        number: core::cmp::min(MAX_PAGE_NUMBER, self.number.saturating_add(rhs)),
+                        number: core::cmp::min(MAX_PAGE_NUMBER, self.number.saturating_add(rhs * P::CONTAINED_4K)),
+                        size: self.size,
                     }
                 }
             }
-            impl AddAssign<usize> for $TypeName {
+            impl<P: PageSize> AddAssign<usize> for $TypeName<P> {
                 fn add_assign(&mut self, rhs: usize) {
                     *self = $TypeName {
-                        number: core::cmp::min(MAX_PAGE_NUMBER, self.number.saturating_add(rhs)),
+                        number: core::cmp::min(MAX_PAGE_NUMBER, self.number.saturating_add(rhs * P::CONTAINED_4K)),
```

Comment:
Not sure why I didn't do this originally, but in `AddAssign` and `SubAssign` we should be able to re-use the logic from `Add` and `Sub` respectively.

For `Add`, something like:
```rust
    *self = (*self).add(rhs);
```